docarray/array/mixins/find.py

```diff
@@ -0,0 +1,268 @@
+from typing import overload, Optional, Union, Dict, Tuple, Callable, TYPE_CHECKING
+from ...math.helper import top_k, minmax_normalize, update_rows_x_mat_best
+from ...score import NamedScore
+import numpy as np
+
+
+if TYPE_CHECKING:
+    from ...types import T, ArrayType
+
+    from ... import Document, DocumentArray
+
+
+class FindMixin:
+    """ A mixin that provides find functionality to DocumentArrays  """
+
+    @overload
+    def find(self: 'T', query: 'ArrayType', **kwargs):
+        ...
+
+    @overload
+    def find(self: 'T', query: Union['Document', 'DocumentArray'], **kwargs):
+        ...
+
+    @overload
+    def find(self: 'T', query: Dict, **kwargs):
+        ...
+
+    @overload
+    def find(self: 'T', query: str, **kwargs):
+        ...
+
+    def find(
+        self: 'T',
+        query: Union['DocumentArray', 'Document', 'ArrayType', Dict, str],
+        **kwargs,
+    ) -> 'DocumentArray':
+        from ...math.ndarray import to_numpy_array, get_array_type
+        from ... import Document, DocumentArray
+
+        if isinstance(query, (dict, str)):
+            return self.filter(query, **kwargs)
+        elif isinstance(query, (DocumentArray, Document)):
+            if isinstance(query, Document):
+                query = DocumentArray(query)
+                return self.search(query, **kwargs)[0].matches
+
+            return self.search(query, **kwargs)
```

Comment:
I think we should either return a document array of matches or a list of match document arrays, not root documents

Reply:
Addressed